lock-api/src/main/java/com/palantir/lock/client/PartitioningTimelockService.java

```diff
@@ -0,0 +1,172 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import static java.util.stream.Collectors.toSet;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Ordering;
+import com.google.common.collect.Streams;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.AutoDelegate_TimelockService;
+import com.palantir.lock.v2.ImmutableLockRequest;
+import com.palantir.lock.v2.LockRequest;
+import com.palantir.lock.v2.LockResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimelockService;
+
+public final class PartitioningTimelockService implements AutoDelegate_TimelockService {
+    private static final int LOCK_BATCH_THRESHOLD = 1000;
+    private static final int UNLOCK_BATCH_THRESHOLD = 10_000;
+
+    private final Clock clock;
+    private final TimelockService timelockService;
+
+    // tokens used as identifiers; these don't necessarily directly exist in timelock
+    private final ConcurrentMap<LockToken, Set<LockToken>> proxyLockTokens = new ConcurrentHashMap<>();
+
+    @VisibleForTesting
+    PartitioningTimelockService(Clock clock, TimelockService timelockService) {
+        this.clock = clock;
+        this.timelockService = timelockService;
+    }
+
+    public PartitioningTimelockService(TimelockService timelockService) {
+        this(Clock.systemUTC(), timelockService);
+    }
+
+    @Override
+    public TimelockService delegate() {
+        return timelockService;
+    }
+
+    private Set<LockToken> addProxies(Set<LockToken> maybeProxies) {
+        if (proxyLockTokens.isEmpty()) {
+            return maybeProxies;
+        } else {
+            return maybeProxies.stream()
+                    .flatMap(token -> Optional.ofNullable(proxyLockTokens.remove(token))
+                            .orElse(Collections.singleton(token)).stream())
+                    .collect(toSet());
+        }
+    }
+
+    @Override
+    public Set<LockToken> unlock(Set<LockToken> unenrichedTokens) {
+        Set<LockToken> tokens = addProxies(unenrichedTokens);
+        if (tokens.size() <= UNLOCK_BATCH_THRESHOLD) {
+            return timelockService.unlock(tokens);
+        }
+        return Streams.stream(Iterables.partition(tokens, UNLOCK_BATCH_THRESHOLD))
+                .map(ImmutableSet::copyOf)
+                .map(timelockService::unlock)
+                .flatMap(Collection::stream)
+                .collect(toSet());
```

Comment:
the result of this contains the proxy tokens and the tokens they proxy, but should probably only contain the proxy tokens.

Reply:
yep makes sense